helpers/response-validator-helper.js

```diff
@@ -0,0 +1,74 @@
+import {fail, check} from "k6";
+
+export class ResponseValidatorHelper {
+    validateResponseStatus(response, expectedStatus, requestName = null) {
+        const validateMessage = requestName
+            ? `Verify that ${requestName} response status is ${expectedStatus}.`
+            : `Verify response status is ${expectedStatus}.`;
+        const failMessage = requestName
+            ? `${requestName} response status is not ${expectedStatus} but ${response.status}.`
+            : `Response status is not ${expectedStatus} but ${response.status}.`;
+
+        if (
+            !check(response, {
+                [validateMessage]: (r) => r.status === expectedStatus,
+        })) {
+            fail(failMessage);
+        }
+    }
+
+    validateSingleResourceResponseJson(responseJson, requestName = null) {
+        const validateDataMessage = requestName
+            ? `Verify ${requestName} response body has 'data' defined.`
+            : `Verify response body has 'data' defined.`;
+        const validateDataAttributesMessage = requestName
+            ? `Verify ${requestName} response body has 'data.attributes' defined.`
+            : `Verify response body has 'data.attributes' defined.`;
+        const failMessage = requestName
+            ? `${requestName} response body does not contain valid JSON.`
+            : `Response body does not contain valid JSON.`;
+
+        if (
+            !check(responseJson, {
+                [validateDataMessage]: (responseJson) => responseJson.data !== undefined,
+                [validateDataAttributesMessage]: (responseJson) => responseJson.data.attributes !== undefined
+            })
+        ) {
+            fail(failMessage);
+        }
+    }
+
+    validateResourceCollectionResponseJson(responseJson, requestName = null) {
+        const validateDataMessage = requestName
+            ? `Verify ${requestName} response body has 'data' defined.`
+            : `Verify response body has 'data' defined.`;
+        const failMessage = requestName
+            ? `${requestName} response body does not contain valid JSON.`
+            : `Response body does not contain valid JSON.`;
+
+        if (
+            !check(responseJson, {
+                [validateDataMessage]: (responseJson) => responseJson.data !== undefined,
+            })
+        ) {
+            fail(failMessage);
```

Comment:
Ok, I agree with your statement, still, I think that in `helper` classes there's no point in continuing the test if you failed to log in or create a cart. So, I suggest we use this `ResponseValidatorHelper` only in other helpers and leave assertions in `AbstractScenario` as is. This way we can decide for each scenario when we should stop the test execution on a failed assertion.

Reply:
I see this as just a convenience wrapper for API response checks that can fail. For example, if you check the response status to be 200 but you get a 500, the server is probably not responding to this request, and the reason could be very well that this request could not be handled while the next can. fail() will interrupt the test completely.

For the login helper, it _might_ make sense, but even there I'm not sure. We will break because one of n logins failed and the next URLs won't even be tried. I just don't know how this will affect the statistics for each URL at the end. 🤔

Reply:
I have looked at this page https://k6.io/docs/using-k6/thresholds I think thresholds are the right way to configure this. As I mentioned before fail() will trigger an exception that is also printed in the console.

I think this is globally configurable and can be overridden for a scenario if needed, so this should be easily doable for all tests.

```ts
export const options = {
  thresholds: {
    http_req_failed: ['rate<1'], // Ideally we want 0% fails for any http_request for the performance tests.
    checks: ['p(100)<1'], // I'm not sure how to check for 100% success
  },
};
```

1. Rename the ResponseValidatorHelper helper to ResponseAssertionsHelper. I'm in favor of omitting the "Helper" suffix, it adds IMHO no additional benefit, but feel free to keep it.
2. Remove the fail()
3. Use thresholds

Reply:
OK, I've added the thresholds. Notice the checks in the summary. I think this is indeed how this should be checked correctly.

Reply:
1. Moved all assertions to a dedicated class `AssertionsHelper`. I dropped the `Response` prefix, because we have assertions not only for responses, and left the `Helper` suffix for consistency :)
2. Removed all `fail()` calls from assertions.